lib/classifier-reborn/extensions/hasher.rb

```diff
@@ -2,38 +2,31 @@
 # Copyright:: Copyright (c) 2005 Lucas Carlson
 # License::   LGPL
 
+require 'set'
+
 module ClassifierReborn
   module Hasher
     STOPWORDS_PATH = [File.expand_path(File.dirname(__FILE__) + '/../../../data/stopwords')]
 
     extend self
 
-    # Removes common punctuation symbols, returning a new string.
-    # E.g.,
-    #   "Hello (greeting's), with {braces} < >...?".without_punctuation
-    #   => "Hello  greetings   with  braces         "
-    def without_punctuation(str)
-      str .tr( ',?.!;:"@#$%^&*()_=+[]{}\|<>/`~', " " ) .tr( "'\-", "")
-    end
-
     # Return a Hash of strings => ints. Each word in the string is stemmed,
     # interned, and indexes to its frequency in the document.
     def word_hash(str, language = 'en')
-      word_hash   = clean_word_hash(str, language)
-      symbol_hash = word_hash_for_symbols(str.gsub(/[\w]/," ").split)
-      return clean_word_hash(str, language).merge(symbol_hash)
+      cleaned_word_hash = clean_word_hash(str, language)
+      symbol_hash = word_hash_for_symbols(str.scan(/[^\s\p{WORD}]/))
+      return cleaned_word_hash.merge(symbol_hash)
     end
 
     # Return a word hash without extra punctuation or short symbols, just stemmed words
     def clean_word_hash(str, language = 'en')
-      word_hash_for_words str.gsub(/[^\w\s]/,"").split, language
+      word_hash_for_words str.gsub(/[^\p{WORD}\s]/,'').downcase.split, language
     end
```

Comment:
What side effects? Its not a downcase or gsub in place, it makes a new object which is split, then passed to word_hash_for_words then downcased. All I do is downcase the entire thing at once instead of word*N times.